lib/1.4/dml-builtins.dml

```diff
@@ -2825,136 +2826,89 @@ template register is (_conf_attribute, get, set, shown_desc,
         local int n = 0;
-        // TODO: it is inefficient to re-generate the field list each call
-        foreach f in (this.fields) {
+        local uint64 lsbs = 0;
+
+        // counting sort
+        foreach f in (_get_fields) {
             local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
+            lsbs[lsb] = 1;
             unmapped[lsb + f.bitsize - 1 : lsb] = 0;
-            ++n;
+            n += 1;
         }
-        assert n <= 64;
-        return (n, unmapped);
-    }
 
-    // fills in the passed array with all fields in this register, and returns
-    // the number of elements and a bitmask showing bits not covered by fields.
-    shared method _get_get_fields(_get_field *fields) -> (int, uint64) /* n, unmapped */ {
-        local uint64 unmapped;
-        unmapped[this.bitsize-1:0] = -1;
-        local int n = 0;
-        // TODO: it is inefficient to re-generate the field list each call
-        foreach f in (this._get_fields) {
-            local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
-            unmapped[lsb + f.bitsize - 1 : lsb] = 0;
-            ++n;
+        foreach f in (_get_fields) {
+            fields[bit_count64(lsbs[f.lsb : 0]) - 1] = f;
         }
-        assert n <= 64;
         return (n, unmapped);
     }
 
     // fills in the passed array with all fields in this register, and returns
     // the number of elements and a bitmask showing bits not covered by fields.
     shared method _get_set_fields(_set_field *fields) -> (int, uint64) /* n, unmapped */ {
         local uint64 unmapped;
-        unmapped[this.bitsize-1:0] = -1;
+        unmapped[this.bitsize - 1 : 0 ] = -1;
         local int n = 0;
-        // TODO: it is inefficient to re-generate the field list each call
-        foreach f in (this._set_fields) {
+        local uint64 lsbs = 0;
+
+        // counting sort
+        foreach f in (_set_fields) {
             local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
+            lsbs[lsb] = 1;
```

Comment:
note the microoptimization:
```C
lsbs |= (1 << lsb);
```
I'm actually going to speak in favor of it since the entire point of this PR is to optimize, and it doesn't really change anything in clarity.

Reply:
You can, however, disregard it if it makes no difference on your benches.

Reply:
Wow! Turns out i don't understand crap and compilers do whatever the hell they want!